src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs

```diff
@@ -376,18 +383,98 @@ internal void DoGetExtensionMethods(ArrayBuilder<MethodSymbol> methods, string n
         }
 
 #nullable enable
+        private static bool IsValidExtensionReceiverParameter(ParameterSymbol thisParam)
+        {
+            Debug.Assert(thisParam is not null);
+
+            // For ref and ref-readonly extension members and classic extension methods, receivers need to be of the correct types to be considered in lookup
+            if (thisParam.RefKind == RefKind.Ref && !thisParam.Type.IsValueType)
+            {
+                return false;
+            }
 
-        internal void GetExtensionContainers(ArrayBuilder<NamedTypeSymbol> extensions)
+            if (thisParam.RefKind is RefKind.In or RefKind.RefReadOnlyParameter && thisParam.Type.TypeKind != TypeKind.Struct)
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        internal void GetExtensionMembers(ArrayBuilder<Symbol> members, string? name, string? alternativeName, int arity, LookupOptions options)
         {
+            Debug.Assert((options & ~(LookupOptions.IncludeExtensionMembers | LookupOptions.AllMethodsOnArityZero
+                | LookupOptions.MustBeInstance | LookupOptions.MustNotBeInstance | LookupOptions.MustBeInvocableIfMember
+                | LookupOptions.MustBeOperator)) == 0);
+
+            Debug.Assert(name is not null || alternativeName is null);
+
             if (!this.IsClassType() || !IsStatic || IsGenericType || !MightContainExtensionMethods) return;
 
-            foreach (var nestedType in GetTypeMembersUnordered())
+            foreach (NamedTypeSymbol nestedType in GetTypeMembersUnordered())
             {
-                if (nestedType.IsExtension)
+                if (nestedType is not { IsExtension: true, ExtensionParameter: { } extensionParameter }
+                    || !IsValidExtensionReceiverParameter(extensionParameter))
                 {
-                    extensions.Add(nestedType);
+                    continue;
+                }
+
+                var candidates = name is null || alternativeName is not null
+                    ? nestedType.GetMembersUnordered()
+                    : nestedType.GetMembers(name);
+
+                foreach (var candidate in candidates)
+                {
+                    if (extensionMemberMatches(candidate, name, alternativeName, arity, options))
```

Comment:
>if (extensionMemberMatches(candidate, name, alternativeName, arity, options)) [](http://example.com/codeflow?start=20&length=77)

Should we be filtering out types?

Reply:
I don't think we need an extra filter. There's no lookup option to exclude types. So it's up to the caller to either request an operator, an invocable member, or deal with various members we could find.